astropy/units/format/generic.py

```diff
@@ -592,6 +600,37 @@ def _do_parse(cls, s, debug=False):
                 else:
                     raise ValueError(f"Syntax error parsing unit '{s}'")
 
+    @classmethod
```

Comment:
I've been looking at the format classes for the speed-up stuff, and one disadvantage of moving this method and the other one up here, is that they actually *fail* on `Generic` itself (since it has no `_units` attribute). That seems rather bad form. We may need an intermediate class, or think a bit more generally on how to refactor (perhaps not worrying about breaking backward compatibility for private methods).

Reply:
We have, over the last few months, found quite a lot of duplicated, dead or buggy code in the unit formatters. To me it seems that making small incremental changes to improve the situation has been the correct strategy even if a few individual updates have turned out to be suboptimal in hindsight. Regarding this change in particular, I will point out that at the time the `_units`, `_deprecated_units` and `_functions` class properties in `Generic` all depended on `_all_units`, which in turn depended on `_generate_unit_names()`, which was not implemented in `Generic`, not even as an abstract method. This means that even before this pull request there was already code in `Generic` that only its subclasses could use, so adding a couple more such methods was not a large qualitative change.

Given that the `astropy` units formatters are implemented as classes, it does make sense to use inheritance to reduce code duplication. But it is less clear if the base class for the parsing formatters should be a proper parsing formatter itself, not an abstract base class. Furthermore, the base class we have (i.e. `Generic`) has the only parser that is not limited to parsing ASCII, so the other formatters inherit quite a lot of Unicode-handling machinery that they don't use. It appears to me that the code would be simplified if the base class for parsing formatters would be less like `Generic` and more like `FITS`. That would be an API change, but major releases are the least bad time for such changes, so we might want to think about it.

Reply:
Fair point that the methods didn't even work before! I very much agree with your choice of going for small, incremental changes to clean things up. But perhaps we are indeed reaching the limit of that - it seems also a bit odd that classes overwrite the parser, which presumably is the main thing that defines them.

Anyway, I don't really have an immediate solution, mostly just pointed this out since I think we are at a stage where we may need to define what we work towards. A possible solution also might be to actually define `_units` on `Generic` - after all, it is basically the registry.

Reply:
> I very much agree with your choice of going for small, incremental changes to clean things up. But perhaps we are indeed reaching the limit of that...

The idea that making small, incremental changes to the code makes the code quality asymptotically approach a local maximum sounds reasonable, as does the idea that this local maximum can differ from the global maximum. But the unit formatters are still quite far away from even the local maximum. I still see more opportunities for cleanup, but I'm avoiding opening multiple pull requests at a time because I want to avoid creating merge conflicts for myself.

Reply:
Indeed! And happy to continue reviewing them!